hazelcast/src/main/java/com/hazelcast/spi/impl/merge/BaseSplitBrainHandlerService.java

```diff
@@ -0,0 +1,158 @@
+ */
+@SuppressWarnings("WeakerAccess")
+public abstract class BaseSplitBrainHandlerService<Store> implements SplitBrainHandlerService {
+
+    private final IPartitionService partitionService;
+    private final OperationExecutor operationExecutor;
+
+    protected BaseSplitBrainHandlerService(NodeEngine nodeEngine) {
+        this.partitionService = nodeEngine.getPartitionService();
+        this.operationExecutor = ((OperationServiceImpl) nodeEngine.getOperationService()).getOperationExecutor();
+    }
+
+    @Override
+    public final Runnable prepareMergeRunnable() {
+        ConcurrentLinkedQueue<Store> mergingStores = new ConcurrentLinkedQueue<Store>();
+
+        collectStores(mergingStores);
+
+        return newMergeRunnable(mergingStores, this);
+    }
+
+    private void collectStores(final ConcurrentLinkedQueue<Store> mergingStores) {
+        int partitionCount = partitionService.getPartitionCount();
+        final CountDownLatch latch = new CountDownLatch(partitionCount);
+
+        for (int i = 0; i < partitionCount; i++) {
+            operationExecutor.execute(new StoreCollector(mergingStores, i, latch));
+        }
+
+        try {
+            latch.await();
+        } catch (InterruptedException e) {
+            currentThread().interrupt();
+        }
+    }
+
+    private class StoreCollector implements PartitionSpecificRunnable {
+        private final int partitionId;
+        private final CountDownLatch latch;
+        private final ConcurrentLinkedQueue<Store> mergingStores;
+
+        public StoreCollector(ConcurrentLinkedQueue<Store> mergingStores,
+                              int partitionId,
+                              CountDownLatch latch) {
+            this.mergingStores = mergingStores;
+            this.partitionId = partitionId;
+            this.latch = latch;
+        }
+
+        @Override
+        public int getPartitionId() {
+            return partitionId;
+        }
+
+        @Override
+        public void run() {
+            LinkedList<Store> storesToDestroy = new LinkedList<Store>();
+            try {
+                Iterator<Store> iterator = storeIterator(partitionId);
+                while (iterator.hasNext()) {
+                    Store store = iterator.next();
+
+                    if (isLocalPartition(partitionId)
+                            && hasEntry(store)
+                            && !hasDiscardPolicy(store)) {
+                        mergingStores.add(store);
+                    } else {
+                        storesToDestroy.add(store);
+                    }
+
+                    iterator.remove();
+                }
+
+                asyncDestroyStores(storesToDestroy, partitionId);
+
+            } finally {
+                latch.countDown();
+            }
+        }
+    }
+
+    void asyncDestroyStores(final Collection<Store> stores, final int partitionID) {
+        operationExecutor.execute(new PartitionSpecificRunnable() {
```

Comment:
Can the partition be migrating? Do we need to retry? See - https://github.com/hazelcast/hazelcast/pull/12691#discussion_r178733516

Reply:
No need to retry, `PartitionSpecificRunnable ` can independently run on local without doing any checks like `is-migration-in-progress`.